contracts/core/Allo.sol

```diff
@@ -33,57 +57,108 @@ contract Allo is Initializable {
+        // If so -> ETH / ERC20 ?
+        // Should we also accept amount as an argument to check ?
+
+        pools[_poolIndex] = pool;
+
+        return _poolIndex++;
     }
 
-    // passes _data & msg.sender through to the allocation strategy for that pool
-    // it should return a uint that represents the application
+    /// @notice passes _data through to the allocation strategy for that pool
+    /// @param _poolId id of the pool
+    /// @param _data encoded data unique to the allocation strategy for that pool
     function applyToPool(
         uint _poolId,
         bytes memory _data
-    ) external payable returns (uint) {
-        // Implement the function here
+    ) external payable returns (uint applicationId) {
+        IAllocationStrategy allocationStrategy = pools[_poolId].allocationStrategy;
+        applicationId = allocationStrategy.applyToRound(_data, msg.sender);
     }
 
-    // decode the _data into what's relevant for this strategy
-    // perform whatever actions are necessary (token transfers, storage updates, etc)
-    // all approvals, checks, etc all happen within internal functions from here?
-    function updateMetadata(
+    /// @notice Update pool metadata
+    /// @param _poolId id of the pool
+    /// @param _metadata new metadata of the pool
+    /// @dev invoked only by pool owner
+    function updatePoolMetadata(
         uint _poolId,
-        bytes memory _data
+        bytes memory _metadata
     ) external payable returns (bytes memory) {
-        // Implement the function here
+        if (!registry.isMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        Pool storage pool = pools[_poolId];
+        pool.metadata = _metadata;
+
+        emit PoolMetadataUpdated(_poolId, _metadata);
     }
 
     // transfers _poolAmt from msg.sender to the pool, and takes a fee
     function fundPool(uint _poolId, uint _poolAmt) external payable {
         // Implement the function here
     }
 
-    // passes _data & msg.sender through to the allocation strategy for that pool
+    /// @notice passes _data & msg.sender through to the allocation strategy for that pool
+    /// @param _poolId id of the pool
+    /// @param _data encoded data unique to the allocation strategy for that pool
     function allocate(uint _poolId, bytes memory _data) external payable {
-        // Implement the function here
+        pools[_poolId].allocationStrategy.allocate(_data, msg.sender);
     }
 
     // calls voting.generatePayouts() and then uses return data for payout.activatePayouts()
     // permissionless for anyone to call, checks happen within the strategies
     // check to make sure they haven't skrited around fee
-    function finalize(uint _poolId) external {
-        // Implement the function here
+    function finalize(uint _poolId, bytes memory _data) external {
+
+        // ASK: Do we need _data to allow owner to pass custom data ?
```

Comment:
@zobront not sure if we are still intending to use finalize
If yes -> then do we allow custom data if a distribution wants to use something that is not given by  generatePayouts?